fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java

```diff
@@ -663,4 +673,105 @@ public LogicalPlan visitTableFunction(TableFunctionRelation node, ExpressionMapp
                 new ExpressionMapping(new Scope(RelationId.of(node), node.getRelationFields()), outputColumns)),
                 null, null);
     }
+
+    private Pair<OptExprBuilder, OptExprBuilder> processOnClauseSubquery(JoinRelation node, OptExprBuilder leftOpt,
+                                                                         OptExprBuilder rightOpt,
+                                                                         ExpressionMapping expressionMapping) {
+        Subquery subquery = node.getOnPredicate().getSubquery();
+        if (subquery == null) {
+            return Pair.create(leftOpt, rightOpt);
+        }
+        QueryStatement subqueryStmt = subquery.getQueryStatement();
+        SelectRelation selectRelation = (SelectRelation) subqueryStmt.getQueryRelation();
+        RelationId subqueryRelationId = selectRelation.getRelation().getScope().getRelationId();
+        List<FieldId> correlatedFieldIds = selectRelation.getColumnReferences().values().stream()
+                .filter(field -> !Objects.equals(subqueryRelationId, field.getRelationId()))
+                .collect(Collectors.toList());
+
+        /*
+         * Apply comprises two children, R and E(r) respectively
+         *         ApplyOperator
+         *       /              \
+         *   Outer:R        Inner: E(r)
+         * Since join node has two relation, we should to determine which one(left or right or both)
+         * is the outer relation of ApplyOperator
+         * kind == LEFT, then the left relation of join will be the outer relation of apply
+         * kind == RIGHT, then the right relation of join will be the outer relation of apply
+         * TODO, kind == BOTH, both of the left and right relations should be taken into account
+         */
+        final int LEFT = 1;
+        final int RIGHT = 2;
+        final int BOTH = 3;
+        final int kind;
+        SubqueryTransformer subqueryTransformer = new SubqueryTransformer(session);
+        if (correlatedFieldIds.isEmpty()) {
+            // Find the smallest part of join on predicate that contains the subquery
+            Expr subqueryConjunct = node.getOnPredicate();
```

Comment:
Abstract  this logic to a special method.

Reply:
I also aggree abstract this logic to a special method. May be subquery related class

Reply:
Done